tests/core/pyspec/eth2spec/test/phase0/fork_choice/test_on_block.py

```diff
@@ -0,0 +1,691 @@
+
+    yield 'steps', test_steps
+
+
+@with_all_phases
+@spec_state_test
+@with_presets([MINIMAL], reason="mainnet config requires too many pre-generated public/private keys")
+def test_on_block_update_justified_checkpoint_within_safe_slots(spec, state):
+    """
+    Test `should_update_justified_checkpoint`:
+    compute_slots_since_epoch_start(get_current_slot(store)) < SAFE_SLOTS_TO_UPDATE_JUSTIFIED
+    """
+    test_steps = []
+    # Initialization
+    store, anchor_block = get_genesis_forkchoice_store_and_block(spec, state)
+    yield 'anchor_state', state
+    yield 'anchor_block', anchor_block
+    current_time = state.slot * spec.config.SECONDS_PER_SLOT + store.genesis_time
+    on_tick_and_append_step(spec, store, current_time, test_steps)
+    assert store.time == current_time
+
+    # Skip epoch 0 & 1
+    for _ in range(2):
+        next_epoch(spec, state)
+    # Fill epoch 2
+    state, store, _ = yield from apply_next_epoch_with_attestations(
+        spec, state, store, True, False, test_steps=test_steps)
+    assert state.finalized_checkpoint.epoch == store.finalized_checkpoint.epoch == 0
+    assert state.current_justified_checkpoint.epoch == store.justified_checkpoint.epoch == 2
+    # Skip epoch 3 & 4
+    for _ in range(2):
+        next_epoch(spec, state)
+    # Epoch 5: Attest current epoch
+    state, store, _ = yield from apply_next_epoch_with_attestations(
+        spec, state, store, True, False, participation_fn=_drop_random_one_third, test_steps=test_steps)
+    assert state.finalized_checkpoint.epoch == store.finalized_checkpoint.epoch == 0
+    assert state.current_justified_checkpoint.epoch == 2
+    assert store.justified_checkpoint.epoch == 2
+    assert state.current_justified_checkpoint == store.justified_checkpoint
+
+    # Skip epoch 6
+    next_epoch(spec, state)
+
+    pre_state = state.copy()
+
+    # Build a block to justify epoch 5
+    signed_block = state_transition_with_full_block(spec, state, True, True)
+    assert state.finalized_checkpoint.epoch == 0
+    assert state.current_justified_checkpoint.epoch == 5
+    assert state.current_justified_checkpoint.epoch > store.justified_checkpoint.epoch
+    assert spec.get_current_slot(store) % spec.SLOTS_PER_EPOCH < spec.SAFE_SLOTS_TO_UPDATE_JUSTIFIED
+    # Run on_block
+    yield from tick_and_add_block(spec, store, signed_block, test_steps)
+    # Ensure justified_checkpoint has been changed but finality is unchanged
+    assert store.justified_checkpoint.epoch == 5
+    assert store.justified_checkpoint == state.current_justified_checkpoint
+    assert store.finalized_checkpoint.epoch == pre_state.finalized_checkpoint.epoch == 0
+
+    yield 'steps', test_steps
+
+
+@with_all_phases
+@with_presets([MINIMAL], reason="It assumes that `MAX_ATTESTATIONS` >= 2/3 attestations of an epoch")
+@spec_state_test
+def test_on_block_outside_safe_slots_but_finality(spec, state):
```

Comment:
@djrtwo 
I largely rewrote this test case. Do you mind taking a look if it matches the case you wanted to test?

Reply:
yeah, qiute a case! but looks correct